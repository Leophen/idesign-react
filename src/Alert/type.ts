export interface AlertProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 是否显示关闭按钮
   * @default false
   */
  closable?: boolean;
  /**
   * 提示内容
   */
  message?: React.ReactNode;
  /**
   * 跟在提示内容后的操作区
   */
  operation?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 标题
   */
  title?: React.ReactNode;
  /**
   * 提示类型
   * @default info
   */
  type?: 'info' | 'success' | 'warning' | 'error';
}
