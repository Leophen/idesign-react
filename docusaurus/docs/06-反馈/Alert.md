---
title: Alert 警示
---

警告提示，展现需要关注的信息。

## 引入

```js
import {Alert} from 'idesign';
```

## 基本用法

最简单的用法，适用于简短的警告提示。

```jsx live fffx
<Alert message="提示成功的文案" type="success" />
```

## 不同样式

共有四种样式 `success`、`info`、`warning`、`error`

```jsx live fffx
<div className="demo-column">
    <Alert message="提示成功的文案" type="success" />
    <Alert message="提示通知的文案" type="info" />
    <Alert message="提示警告的文案" type="warning" />
    <Alert message="提示报错的文案" type="error" />
</div>
```

## 带标题

带标题含有辅助性文字介绍的警告提示。

```jsx live fffx
<div className="demo-column">
    <Alert
        message="成功标题"
        description="提示成功的文案"
        type="success"
    />
    <Alert
        message="通知标题"
        description="提示通知的文案"
        type="info"
    />
    <Alert
        message="警告标题"
        description="提示警告的文案"
        type="warning"
    />
    <Alert
        message="报错标题"
        description="提示报错的文案"
        type="error"
    />
</div>
```

## 顶部公告

页面顶部通告形式，默认有图标且 `type` 为 `'warning'`

```jsx live fffx
<div className="demo-column">
    <Alert message="提示警告的文案" banner />
    <Alert
        message="提示警告的文案"
        banner
        closable
    />
    <Alert showIcon={false} message="提示警告的文案" banner />
    <Alert type="error" message="提示报错的文案" banner />
</div>
```

## 带图标

图标让信息类型更加醒目。

```jsx live fffx
<div className="demo-column">
    <Alert message="提示成功的文案" type="success" showIcon />
    <Alert message="提示通知的文案" type="info" showIcon />
    <Alert message="提示警告的文案" type="warning" showIcon closable />
    <Alert message="提示报错的文案" type="error" showIcon />
    <Alert
        message="成功标题"
        description="提示成功的文案"
        type="success"
        showIcon
    />
    <Alert
        message="通知标题"
        description="提示通知的文案"
        type="info"
        showIcon
    />
    <Alert
        message="警告标题"
        description="提示警告的文案"
        type="warning"
        showIcon
        closable
    />
    <Alert
        message="报错标题"
        description="提示报错的文案"
        type="error"
        showIcon
    />
</div>
```

## 自定义图标

通过 `icon` 属性自定义图标。

```jsx live fffx
<div className="demo-column">
    <Alert icon={<OutlinedSmile />} message="提示成功的文案" type="success" />
    <Alert icon={<OutlinedSmile />} message="提示成功的文案" type="success" showIcon />
    <Alert icon={<OutlinedSmile />} message="提示通知的文案" type="info" showIcon />
    <Alert icon={<OutlinedSmile />} message="提示警告的文案" type="warning" showIcon />
    <Alert icon={<OutlinedSmile />} message="提示报错的文案" type="error" showIcon />
    <Alert
        icon={<OutlinedSmile />}
        message="成功标题"
        description="提示成功的文案"
        type="success"
        showIcon
    />
    <Alert
        icon={<OutlinedSmile />}
        message="通知标题"
        description="提示通知的文案"
        type="info"
        showIcon
    />
    <Alert
        icon={<OutlinedSmile />}
        message="警告标题"
        description="提示警告的文案"
        type="warning"
        showIcon
    />
    <Alert
        icon={<OutlinedSmile />}
        message="报错标题"
        description="提示报错的文案"
        type="error"
        showIcon
    />
</div>
```

## 可关闭

显示关闭按钮，点击可关闭警告提示。

```jsx live fff
function App() {
    const onClose = (e) => {
        console.log(e, 'I was closed.');
    };

    return (
        <div className="demo-column">
            <Alert
                message="提示警告的文案"
                type="warning"
                closable
                onClose={onClose}
            />
            <Alert
                message="报错标题"
                description="提示报错的文案"
                type="error"
                closable
                onClose={onClose}
            />
        </div>
    );
}
```

## 操作项

可以在右上角自定义操作项。

```jsx live fff
<Alert
    message="提示成功的文案"
    type="success"
    showIcon
    action={
        <Link
            href="http://www.baidu.com"
            target="_blank"
            type="primary"
        >
            超链接
        </Link>
    }
    closable
/>
```

## 平滑地卸载

平滑、自然的卸载提示。

```jsx live fff
function App() {
    const [visible, setVisible] = useState(true);
    const handleClose = () => {
        setVisible(false);
    };

    return (
        <div className="demo-column">
            <p>placeholder text here</p>
            {visible ? (
                <Alert message="提示成功的文案" type="success" closable afterClose={handleClose} />
            ) : null}
        </div>
    );
}
```

## React 错误处理

友好的 React 错误处理 包裹组件。

```jsx live fff
function App() {
    const ErrorBoundary = Alert.ErrorBoundary;
    const ThrowError = () => {
        const [error, setError] = useState();
        const onClick = () => {
            setError(new Error('An Uncaught Error'));
        };

        if (error) {
            throw error;
        }
        return (
            <Button danger onClick={onClick}>
                Click me to throw a error
            </Button>
        );
    };

    return (
        <ErrorBoundary>
            <ThrowError />
        </ErrorBoundary>
    );
}
```

## API

| 参数              | 说明                                                                                               | 类型                      | 默认值       |
| ----------------- | -------------------------------------------------------------------------------------------------- | ------------------------- | ------------ |
| children          | 嵌套的标题                                                                                         | `ReactNode`               | `--`         |
| className         | 分割线样式类                                                                                       | `string`                  | `--`         |
| dashed            | 是否虚线                                                                                           | `boolean`                 | `false`      |
| orientation       | 分割线标题的位置                                                                                   | `left`〡`right`〡`center` | `center`     |
| orientationMargin | 标题和最近 left/right 边框之间的距离，去除了分割线<br/>同时 `orientation` 必须为 `left` 或 `right` | `string`〡`number`        | `--`         |
| plain             | 文字是否显示为普通正文样式                                                                         | `boolean`                 | `false`      |
| style             | 分割线样式对象                                                                                     | `CSSProperties`           | `--`         |
| type              | 水平还是垂直类型                                                                                   | `horizontal`〡`vertical`  | `horizontal` |

