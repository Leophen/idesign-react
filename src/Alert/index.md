---
nav:
  title: 组件
  path: /components
group:
  title: 弹框提示组件
  order: 5
order: 1
---

# Alert 提示

用于页面中展示重要的提示信息。

## 基本用法

Alert 组件提供了四种不同的提示类型：`info`（默认）、`success`、`warning` 和 `error`。

```tsx
import React from 'react';
import { Alert } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Alert type="info" message="这是一条普通的消息提示" />
      <Alert type="success" message="这是一条成功的消息提示" />
      <Alert type="warning" message="这是一条警示消息" />
      <Alert type="error" message="高危操作 / 出错信息提示" />
    </div>
  );
};

export default App;
```

## 带关闭按钮的提示

提供 `closable` 属性控制提示是否可关闭。

```tsx
import React from 'react';
import { Alert } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Alert type="info" message="这是一条普通的消息提示" closable />
      <Alert type="success" message="这是一条成功的消息提示" closable />
      <Alert type="warning" message="这是一条警示消息" closable />
      <Alert type="error" message="高危操作 / 出错信息提示" closable />
    </div>
  );
};

export default App;
```

## 带相关操作的提示

提供 `operation` 属性来配置跟在提示内容后的操作。

```tsx
import React from 'react';
import { Alert } from 'idesign-react';

const App = () => {
  const operation = <span>相关操作</span>;

  return (
    <div className="idesign-demo-block-column">
      <Alert type="info" message="这是一条普通的消息提示" operation={operation} closable />
      <Alert type="success" message="这是一条成功的消息提示" operation={operation} closable />
      <Alert type="warning" message="这是一条警示消息" operation={operation} closable />
      <Alert type="error" message="高危操作 / 出错信息提示" operation={operation} closable />
    </div>
  );
};

export default App;
```

## 带标题的提示

提供 `title` 属性来配置提示内容的标题。

```tsx
import React from 'react';
import { Alert } from 'idesign-react';

const App = () => {
  const operation = <span>相关操作</span>;

  return (
    <div className="idesign-demo-block-column">
      <Alert
        type="info"
        title="消息提示标题"
        message="这是一条普通的消息提示"
        operation={operation}
        closable
      />
      <Alert
        type="success"
        title="消息提示标题"
        message="这是一条成功的消息提示"
        operation={operation}
        closable
      />
      <Alert
        type="warning"
        title="消息提示标题"
        message="这是一条警示消息"
        operation={operation}
        closable
      />
      <Alert
        type="error"
        title="消息提示标题"
        message="高危操作 / 出错信息提示"
        operation={operation}
        closable
      />
    </div>
  );
};

export default App;
```

<API />
