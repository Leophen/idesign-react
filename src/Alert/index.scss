.i-alert {
  position: relative;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  width: 100%;
  margin: 0;
  padding: 12px 20px;
  overflow: hidden;
  border-radius: 4px;
  opacity: 1;
  transition: opacity 0.2s ease-out;
}

.i-alert--content {
  padding-right: 26px;
}

.i-alert--title {
  margin-bottom: 6px;
  font-weight: bold;
  font-size: 14px;
}

.i-alert--description {
  box-sizing: border-box;
  margin: 0;
  font-size: 14px;
  word-break: break-all;
}

.i-alert--operation {
  display: inline-block;
  box-sizing: border-box;
  margin-left: 8px;
  color: var(--i-primary-txt);
  cursor: pointer;
  user-select: none;
  &:hover {
    color: var(--i-primary-hover);
  }
  &:active {
    color: var(--i-primary-active);
  }
}

.i-alert--close-btn {
  position: absolute;
  right: 16px;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 4px;
  cursor: pointer;
  &:hover {
    .icon-Close {
      color: var(--i-font-3) !important;
      transition: color 0.2s ease-out;
    }
  }
}

@mixin injectColor($color, $color_background) {
  color: $color;
  background-color: $color_background;
  .i-icon {
    align-self: flex-start;
    margin-top: 2px;
    margin-right: 8px;
    color: $color !important;
  }
  .icon-Close {
    margin-right: 0;
    color: var(--i-font-5) !important;
  }
}

.i-alert--type-info {
  @include injectColor(var(--i-font-5), var(--i-bg-hover));
}

.i-alert--type-success {
  @include injectColor(var(--i-success), var(--i-success-bg));
}

.i-alert--type-warning {
  @include injectColor(var(--i-warning), var(--i-warning-bg));
}

.i-alert--type-error {
  @include injectColor(var(--i-error), var(--i-error-bg));
}
