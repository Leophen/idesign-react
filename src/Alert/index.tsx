import React, { useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import { AlertProps } from './type'

const Alert: React.FC<AlertProps> = (props) => {
  const {
    className,
    closable = false,
    message = '',
    operation,
    style,
    title,
    type = 'info',
    ...restProps
  } = props;

  const [closed, setClosed] = useState(false);

  const handleCloseAlert = () => {
    setClosed(true);
  };

  return closed ? null : (
    <div
      className={classNames('i-alert', `i-alert--type-${type}`, className)}
      style={{ ...style }}
      {...restProps}
    >
      <Icon
        name={
          {
            info: 'TipInfo',
            success: 'TipCheckFill',
            warning: 'TipWarningFill',
            error: 'TipCloseFill',
          }[type]
        }
        size={16}
      />

      <div className="i-alert--content">
        {title && <div className="i-alert--title">{title}</div>}
        <div className="i-alert--description">
          {message}
          {operation && <div className="i-alert--operation">{operation}</div>}
        </div>
      </div>

      {closable && (
        <div className="i-alert--close-btn" onClick={handleCloseAlert}>
          <Icon name="Close" size={16} />
        </div>
      )}
    </div>
  );
};

Alert.displayName = 'Alert';

export default Alert;
