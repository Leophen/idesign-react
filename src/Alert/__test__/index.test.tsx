import { render } from '@testing-library/react';
import Alert from '../index';

describe('Alert 组件测试', () => {
  const testId = 'alert-test-id';

  it('type', () => {
    const props = {
      message: 'Alert 内容',
    };
    const { asFragment } = render(<Alert {...props} type="info" />);
    expect(asFragment()).toMatchSnapshot();

    const { asFragment: asFragment2 } = render(<Alert {...props} type="success" />);
    expect(asFragment2()).toMatchSnapshot();

    const { asFragment: asFragment3 } = render(<Alert {...props} type="warning" />);
    expect(asFragment3()).toMatchSnapshot();

    const { asFragment: asFragment4 } = render(<Alert {...props} type="error" />);
    expect(asFragment4()).toMatchSnapshot();
  });
});

