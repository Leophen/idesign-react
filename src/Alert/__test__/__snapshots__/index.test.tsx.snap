// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert 组件测试 type 1`] = `
<DocumentFragment>
  <div
    class="i-alert i-alert--type-info"
  >
    <i
      class="i-icon icon-TipInfo"
      style="font-size: 16px;"
    />
    <div
      class="i-alert--content"
    >
      <div
        class="i-alert--description"
      >
        Alert 内容
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert 组件测试 type 2`] = `
<DocumentFragment>
  <div
    class="i-alert i-alert--type-success"
  >
    <i
      class="i-icon icon-TipCheckFill"
      style="font-size: 16px;"
    />
    <div
      class="i-alert--content"
    >
      <div
        class="i-alert--description"
      >
        Alert 内容
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert 组件测试 type 3`] = `
<DocumentFragment>
  <div
    class="i-alert i-alert--type-warning"
  >
    <i
      class="i-icon icon-TipWarningFill"
      style="font-size: 16px;"
    />
    <div
      class="i-alert--content"
    >
      <div
        class="i-alert--description"
      >
        Alert 内容
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert 组件测试 type 4`] = `
<DocumentFragment>
  <div
    class="i-alert i-alert--type-error"
  >
    <i
      class="i-icon icon-TipCloseFill"
      style="font-size: 16px;"
    />
    <div
      class="i-alert--content"
    >
      <div
        class="i-alert--description"
      >
        Alert 内容
      </div>
    </div>
  </div>
</DocumentFragment>
`;
